src/node/rpc/frontend.h

```diff
@@ -292,136 +218,40 @@ namespace ccf
+      if (!endpoint->authn_policies.empty())
       {
-        set_response_unauthorized(
-          ctx, fmt::format("'{}' RPC must be signed", ctx->get_method()));
-        update_metrics(ctx, metrics);
-        return ctx->serialise_response();
-      }
-
-      bool should_record_client_signature = false;
-      if (signed_request.has_value())
-      {
-        // For forwarded requests (raft only), skip verification as it is
-        // assumed that the verification was done by the forwarder node.
-        if (
-          (!ctx->is_create_request &&
-           (!(consensus != nullptr &&
-              consensus->type() == ConsensusType::CFT) ||
-            !ctx->session->original_caller.has_value())) &&
-          !verify_client_signature(
-            ctx->session->caller_cert, caller_id, signed_request.value()))
+        std::string auth_error_reason;
+        for (const auto& policy : endpoint->authn_policies)
         {
-          set_response_unauthorized(ctx);
-          update_metrics(ctx, metrics);
-          return ctx->serialise_response();
-        }
-
-        // By default, signed requests are verified and recorded, even on
-        // endpoints that do not require client signatures
-        if (is_primary)
-        {
-          should_record_client_signature = true;
-        }
-      }
-
-      std::optional<Jwt> jwt;
-      if (endpoint->properties.require_jwt_authentication)
-      {
-        auto headers = ctx->get_request_headers();
-        std::string error_reason;
-        auto token = http::JwtVerifier::extract_token(headers, error_reason);
-        auto keys_view = tx.get_view<JwtPublicSigningKeys>(
-          ccf::Tables::JWT_PUBLIC_SIGNING_KEYS);
-        auto key_issuer_view = tx.get_view<JwtPublicSigningKeyIssuer>(
-          ccf::Tables::JWT_PUBLIC_SIGNING_KEY_ISSUER);
-        std::string key_issuer;
-        if (token.has_value())
-        {
-          auto key_id = token.value().header_typed.kid;
-          auto token_key = keys_view->get(key_id);
-          if (!token_key.has_value())
-          {
-            error_reason = "JWT signing key not found.";
-          }
-          else if (!http::JwtVerifier::validate_token_signature(
-                     token.value(), token_key.value()))
-          {
-            error_reason = "JWT signature is invalid.";
-          }
-          else
+          identity = policy->authenticate(tx, ctx, auth_error_reason);
+          if (identity != nullptr)
           {
-            key_issuer = key_issuer_view->get(key_id).value();
+            break;
           }
         }
-        if (!error_reason.empty())
+
+        if (identity == nullptr)
         {
-          set_response_unauthorized_jwt(
-            ctx, fmt::format("'{}' {}", ctx->get_method(), error_reason));
+          // If none were accepted, let the first set an error
+          endpoint->authn_policies[0]->set_unauthenticated_error(
+            ctx, std::move(auth_error_reason));
```

Comment:
This will use the error message of the last policy and pass it to the first policy. Needs extra logic to prevent filling the error message after the first policy.

Reply:
Or just use the last policy :)

Reply:
Made it the last.